src/limb.rs

```diff
@@ -0,0 +1,277 @@
+    // "Key Pair Generation by Testing Candidates".
+    //
+    // TODO: DRY-up with `ec::suite_b::private_key::generate_private_key`?
+    //
+    /// Chooses a positive integer within the range and stores it into `dest`
+    /// encoded as limbs ordered least-significant-limb to most-significant-limb.
+    ///
+    /// This function is intended to be suitable for generating private keys.
+    pub fn sample_into_limbs(&self, dest: &mut [Limb], rng: &rand::SecureRandom)
+                             -> Result<(), ()> {
+
+        assert_eq!(self.max_exclusive.len(), dest.len());
+
+        // Ensure that we have at least a 50% chance of generating a matching
+        // number on each iteration; TODO: Either embed this in the type, or
+        // fix the algorithm so that it is unnecessary
+        debug_assert!(self.min_inclusive < (1 << LIMB_BITS - 1));
+
+        let bits_to_mask = self.leading_zero_bits();
+
+        // XXX: The value 100 was chosen to match OpenSSL due to uncertainty of
+        // what specific value would be better, but it seems bad to try 100 times.
+        for _ in 0..100 {
+            {
+                let mut dest_as_bytes = limbs_as_bytes_mut(dest);
+                try!(rng.fill(&mut dest_as_bytes));
+            }
+
+            if bits_to_mask > 0 {
+                let mask: Limb = (1 << (LIMB_BITS - bits_to_mask)) - 1;
+                dest[self.max_exclusive.len() - 1] &= mask;
+            }
+
+            if !self.are_limbs_within(&dest) {
+                continue;
+            }
+
+            return Ok(());
+        }
+
+        Err(())
+    }
+
+}
+
+#[allow(unsafe_code)]
+#[allow(non_snake_case)]
+#[doc(hidden)]
+#[no_mangle]
+pub unsafe extern fn GFp_rand_range_ex(dest: *mut Limb,
+                                       min_inclusive: Limb,
+                                       max_exclusive: *const Limb,
+                                       num_limbs: c::size_t,
+                                       rng: *mut rand::RAND)
+                                      -> c::int {
+    const ERR: c::int = 0;
+    const SUCCESS: c::int = 1;
+
+    let range = Range::from_bounds_unchecked(
+        min_inclusive,
+        core::slice::from_raw_parts(max_exclusive, num_limbs));
+    let mut dest = core::slice::from_raw_parts_mut(dest, num_limbs);
+
+    let result = range.sample_into_limbs(&mut dest, (*rng).rng);
+    if result.is_err() {
+        return ERR;
+    }
+
+    SUCCESS
+}
+
+
+#[cfg(target_pointer_width = "64")]
+fn limbs_as_bytes_mut<'a>(src: &'a mut [Limb]) -> &'a mut [u8] {
+    polyfill::slice::u64_as_u8_mut(src)
+}
+
+#[cfg(target_pointer_width = "32")]
+fn limbs_as_bytes_mut<'a>(src: &'a mut [Limb]) -> &'a mut [u8] {
+    polyfill::slice::u32_as_u8_mut(src)
+}
+
+// Assumes limbs are ordered least-significant to most-significant, and is not
+// constant time. TODO Is there an easy constant-time implementation?
+fn limbs_gte_limb(limbs: &[Limb], min_inclusive: Limb) -> bool {
```

Comment:
You can replace this with a call to `GFp_constant_time_limbs_are_zero` since the minimum is always 1.